students/ptfbanks/Lesson05/pf_mailroom-3.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+#Instructions:\ Add exceptions and use comprehensions to improve your mailroom project.
+#Exceptions: update your code to handle errors better � like when a user inputs bad data.
+#Comprehensions: Can you use comprehensions to clean up your code a bit?
+#---------------------Set-up----------------------------------#
+#  MailRoom III exercise lesson 05 Include Compression and exception handling
+print('Your desktop has been up-dated with the Donations Management II applicaton.\n'
+'We trust you will appreciate the added menue options and snappy output.\n'
+'select history report option if a list of donors is needed.\n'
+'Try out the new features.\n')
+
+import os #opens access to file creation and placement
+
+#-------Library and definitions-----------------------------------#
+d_dict = {
+'Bob Williams': [5580.00, 3245.50, 1000],
+'Tom Haskell':[345.05, 3245.90, 243.05],
+'Earl Jackson':[4553.45, 45.6, 345.00],
+'Les Thomas':[8934.00, 325.00, 3665.00],
+'James Black':[154, 1300, 21],
+'Ed Jones':[35, 1]}
+
+# experimental course exercise - recorded, but not applied:
+#d_name = {'first_name': 'Bob', 'last_name': 'Williams'}
+#print('My name is {first_name} {last_name}'.format(**d_name))
+
+choice= ("Please select:\n"
+    "     1    to enter a new donation\n"
+    "     2    for a report of donors and gift history\n"
+    "     3    to generate thank you letters, for one or all donors\n"
+    "     4    to quit\n"
+    " your selection:")
+
+letter = ("Dear {},\n"
+   "\t Thank you for your very kind donation(s) totaling $ {}.\n"
+   "It will be put to very good use.\n\n"
+   "\t\t Sincereley,\n"
+   "\t\t   -The Team")
+
+#-------------------------------Functions----------------------#
+
+def hist_rpt():   #Prepare list of donors, sorted by sum donation
+    print("|   DonorName    | TotaL Given | Num Gifts   | Avrage Gift |")
+    print("|----------------|-------------|-------------|-------------|")
+    d_srt = {name: sum(d_dict[name])for name in d_dict}     #Compression for:
+#    d_srt = {}
+#    for name in d_dict:
+#        tot = sum(d_dict[name])
+#        d_srt[name]=tot
+    ordered = sorted(d_srt, key=d_srt.__getitem__, reverse=True)
+    for name in ordered:
+        t_val=d_srt[name]
+        n_o_d=len(d_dict[name])
+        s_avg= t_val/n_o_d
+        print("|{:^16}|{:>13,.02f}|{:^13}|{:>13,.02f}|".format(name, t_val, n_o_d, s_avg))
+        print("|----------------|-------------|-------------|-------------|")        
+    
+def menu_select(choice):  #select and apply menu  (Excepton check incluced)
+    try:
+        while True:
+            pick=input(choice)
+            if pick =='4':
+                print('So long.....')
+                break
+            menu[pick]()
+    except KeyError:
+        print('command not understood.  Plese try again.\n')
+        menu_select(choice)
```

Comment:
your KeyError is supposed to be only around the code that will throw that exception, in this case around line 61.

The main reason why your loop logic is now behaving differently is because you have try/except around the actual while True; which isn't the case almost always.